sys/linux/io_uring.txt

```diff
@@ -3,56 +3,337 @@
+	sq_off		array[const[0, int32], 10]
+	cq_off		array[const[0, int32], 10]
 }
 
-io_sqring_offsets {
-	head		const[0, int32]
-	tail		const[0, int32]
-	ring_mask	const[0, int32]
-	ring_entries	const[0, int32]
-	flags		const[0, int32]
-	dropped		const[0, int32]
-	array		const[0, int32]
-	resv1		const[0, int32]
-	resv2		const[0, int64]
+# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
+# Descriptions for sq_ring and cq_ring manipulation # # # # # # # # # # # # # #
+# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
+
+# Retrieve the cqe at the head of the cq_ring and advance the head. The only meaningful
+# resource contained within a cqe is by the completion of openat or openat2 calls,
+# which produce fd. If that is the case, returns the return value of those. Otherwise,
+# for other operations, returns an invalid fd (-1).
+syz_io_uring_complete(cq_ring_ptr cq_ring_ptr) fd
+
+# Submit sqe into the sq_ring (the submission )
+syz_io_uring_submit(sq_ring_ptr sq_ring_ptr, sqes_ptr sqes_ptr, sqe ptr[in, io_uring_sqe_u], sqes_index int32)
+
+io_uring_sqe_u [
+	IORING_OP_NOP			io_uring_sqe_nop
+	IORING_OP_READV			io_uring_sqe_readv
+	IORING_OP_WRITEV		io_uring_sqe_writev
+	IORING_OP_FSYNC			io_uring_sqe_fsync
+	IORING_OP_READ_FIXED		io_uring_sqe_read_fixed
+	IORING_OP_WRITE_FIXED		io_uring_sqe_write_fixed
+	IORING_OP_POLL_ADD		io_uring_sqe_poll_add
+	IORING_OP_POLL_REMOVE		io_uring_sqe_poll_remove
+	IORING_OP_SYNC_FILE_RANGE	io_uring_sqe_sync_file_range
+	IORING_OP_SENDMSG		io_uring_sqe_sendmsg
+	IORING_OP_RECVMSG		io_uring_sqe_recvmsg
+	IORING_OP_TIMEOUT		io_uring_sqe_timeout
+	IORING_OP_TIMEOUT_REMOVE	io_uring_sqe_timeout_remove
+	IORING_OP_ACCEPT		io_uring_sqe_accept
+	IORING_OP_ASYNC_CANCEL		io_uring_sqe_async_cancel
+	IORING_OP_LINK_TIMEOUT		io_uring_sqe_link_timeout
+	IORING_OP_CONNECT		io_uring_sqe_connect
+	IORING_OP_FALLOCATE		io_uring_sqe_fallocate
+	IORING_OP_OPENAT		io_uring_sqe_openat
+	IORING_OP_CLOSE			io_uring_sqe_close
+	IORING_OP_FILES_UPDATE		io_uring_sqe_files_update
+	IORING_OP_STATX			io_uring_sqe_statx
+	IORING_OP_READ			io_uring_sqe_read
+	IORING_OP_WRITE			io_uring_sqe_write
+	IORING_OP_FADVISE		io_uring_sqe_fadvise
+	IORING_OP_MADVISE		io_uring_sqe_madvise
+	IORING_OP_SEND			io_uring_sqe_send
+	IORING_OP_RECV			io_uring_sqe_recv
+	IORING_OP_OPENAT2		io_uring_sqe_openat2
+	IORING_OP_EPOLL_CTL		io_uring_sqe_epoll_ctl
+	IORING_OP_SPLICE		io_uring_sqe_splice
+	IORING_OP_PROVIDE_BUFFERS	io_uring_sqe_provide_buffers
+	IORING_OP_REMOVE_BUFFERS	io_uring_sqe_remove_buffers
+	IORING_OP_TEE			io_uring_sqe_tee
+]
+
+# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
+# io_uring submission queue entry (io_uring_sqe) descriptions # # # # # # # # #
+# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
+
+#
+# sqe type template
+#
+
+type io_uring_sqe_t[OP, IOPRIO, FD, OFF, ADDR, LEN, MISC_FLAGS, USER_DATA, MISC] {
```

Comment:
There is a useful trick to ensure that all callers don't pass wrong types here (to some degree):

```
type io_uring_sqe_t[OP, IOPRIO, FD, OFF, ADDR, LEN, MISC_FLAGS, USER_DATA, MISC] {
...
} [size[IO_URING_SQE_SIZE]

define IO_URING_SQE_SIZE sizeof(io_uring_sqe)
```

This will ensure that in all cases we at least don't overflow the right size.

Reply:
[added](https://github.com/google/syzkaller/pull/1926/commits/698f056b535e57897347906c27af968a9e5e7543#diff-2e824a90609ef73cc6319dbc406b6b5eR183-R185)